compiler/qsc_eval/src/backend/noise_tests.rs

```diff
@@ -0,0 +1,181 @@
+        .expect("Expected construction of 0.01 depolarizing noise.");
+    assert!(!noise.is_noiseless(), "Expected noise to be noisy.");
+    assert!(
+        0.0 <= noise.distribution[0]
+            && noise.distribution[0] <= noise.distribution[1]
+            && noise.distribution[1] <= noise.distribution[2]
+            && noise.distribution[2] <= 1.1,
+        "Expected non-decreasing noise distribution."
+    );
+    let _ = PauliNoise::from_probabilities(-1e-10, 0.1, 0.1)
+        .expect_err("Expected error for probabilities -1e-10, 0.1, 0.1.");
+    let _ = PauliNoise::from_probabilities(1.0 + -1e-10, 0.1, 0.1)
+        .expect_err("Expected error for probabilities 1.0+1e-10, 0.1, 0.1.");
+    let _ = PauliNoise::from_probabilities(0.3, 0.4, 0.5)
+        .expect_err("Expected error for probabilities 0.3, 0.4, 0.5.");
+}
+
+#[test]
+fn noisy_simulator() {
+    let sim = SparseSim::new();
+    assert!(sim.is_noiseless(), "Expected noiseless simulator.");
+
+    let noise = PauliNoise::from_probabilities(0.0, 0.0, 0.0).expect("Cannot construct noise.");
+    let sim = SparseSim::new_with_noise(&noise);
+    assert!(sim.is_noiseless(), "Expected noiseless simulator.");
+
+    let noise = PauliNoise::from_probabilities(1e-10, 0.0, 0.0).expect("Cannot construct noise.");
+    let sim = SparseSim::new_with_noise(&noise);
+    assert!(!sim.is_noiseless(), "Expected noisy simulator.");
+
+    let noise = PauliNoise::from_probabilities(0.0, 0.0, 1e-10).expect("Cannot construct noise.");
+    let sim = SparseSim::new_with_noise(&noise);
+    assert!(!sim.is_noiseless(), "Expected noisy simulator.");
+}
+
+#[test]
+fn noiseless_gate() {
+    let noise = PauliNoise::from_probabilities(0.0, 0.0, 0.0).expect("Cannot construct noise.");
+    let mut sim = SparseSim::new_with_noise(&noise);
+    let q = sim.qubit_allocate();
+    for _ in 0..100 {
+        sim.x(q);
+        let res1 = sim.m(q);
+        assert!(res1, "Unexpected value. True expected without noise.");
+        sim.x(q);
+        let res2 = sim.m(q);
+        assert!(!res2, "Unexpected value. False expected without noise.");
+    }
+    assert!(sim.qubit_release(q), "Unexpected qubit state on release.");
+}
+
+#[test]
+fn bitflip_measurement() {
+    let noise = PauliNoise::from_probabilities(1.0, 0.0, 0.0).expect("Cannot construct noise.");
+    let mut sim = SparseSim::new_with_noise(&noise);
+    assert!(!sim.is_noiseless(), "Expected noisy simulator.");
+    let q = sim.qubit_allocate();
+    for _ in 0..100 {
+        let res1 = sim.m(q); // Always applies X before measuring
+        assert!(
+            res1,
+            "Unexpected value. True expected for 100% bitflip noise."
+        );
+        let res2 = sim.m(q); // Always applies X before measuring
+        assert!(
+            !res2,
+            "Unexpected value. False expected for 100% bitflip noise."
+        );
+    }
+    assert!(sim.qubit_release(q), "Unexpected qubit state on release.");
```

Comment:
here too, I don't think the loop is really needed.

Reply:
Same as previous comment.